### ISSUE TITLE
What does it mean to be an IL opcode?

### ISSUE THREAD

**Issue by 0xdaryl** [eclipse/omr#5972]
An existential question, but an important one to answer as we review and evaluate the IL opcode set in the compiler and consider adding new ones in the future:  what criteria should we use to define whether some functionality should be implemented as an IL opcode versus some other representation like a function call?

My motivation for asking is driven by a need to better document the IL and by reviewing the existing set of IL opcodes in OMR for the purposes of defining the compiler API.  There will also be a proposal coming forth soon to add better vector support to the compiler which could be introducing 1800 new IL opcodes.  Having a clearer idea in our minds as to what it means to be an IL opcode will help set the design direction there and avoid setting precedents that we don't want.

I'm interested in hearing opinions either in this issue or in a discussion at the April 29 OAM (#5915).

**Comment by 0xdaryl** [comment]
Tagging a few people if they have any opinions: @gita-omr @vijaysun-omr @mstoodle @jdmpapin @fjeremic @joransiu @ymanton @zl-wang @Leonardo2718

**Comment by fjeremic** [comment]
IIRC we talked about this back on another OMR Architecture call where we discussed non-helper symbols and intrinsics. There was some discussion on that call about value to the optimizer, abstraction of a "common" operation, cross platform usage, etc.

**Comment by vijaysun-omr** [comment]
There was a paper attempt that we did some years back, that I had shared with @0xdaryl which got into this kind of a question in quite a lot of detail. Maybe that should be attached for reference here to avoid having to rehash that thought process incrementally.

**Comment by 0xdaryl** [comment]
> IIRC we talked about this back on another OMR Architecture call where we discussed non-helper symbols and intrinsics.

Yeah, I was planning on going back through that discussion at some point too.  I recall that had the perspective of what it means to be an "intrinsic" (or a "non-helper").

**Comment by 0xdaryl** [comment]
Thanks to those who sent me materials offline to help define this.  What I'll do is sort through all that and then roll that into the documentation being produced for the IL for future discussion and review.   I will remove the topic from the OAM today, but I will leave this issue open for a bit longer in case there are any opinions anyone wishes to share.
